doc/dev/kdb-record.md

```diff
@@ -0,0 +1,110 @@
+# `kdb record` Command
+
+The `kdb record` tooling enables the user to record configuration sessions. It also allows exporting and importing the configuration changes.
+
+Recording sessions are stored on a per-user basis. This implies that a recording session can only track changes made by a single user.
+
+## Session handling
+
+* `kdb record start [session_name]`: Starts a new recording session. If no session name is provided, a new random name is generated. If a session with the provided name already exists, changes will be appended.
+* `kdb record stop`: Stops the current recording session. 
+* `kdb record delete [session_name]`: Deletes the specified recording session. If no session name is provided, the current session will be deleted. Deleting a session will not stop the recording. 
+* `kdb record list-sessions`: Lists all sessions by their name
+* `kdb record changes [session_name]`: Shows all changes for the specified session. If no session name is provided, the current session will be used. 
+
+## Exporting configuration changes
+
+`kdb record export [session_name] [--format <format>]`: Will dump the changed keys of the specified session onto `stdout`. If no session name is provided, the current session will be used. 
+
+As example, suppose the key `user:/my/app/hello` has the value `world` and you execute the following commands:
+* `kdb set user:/my/app/hello max`
+* `kdb set user:/my/app/hello susi`
+
+Only the value `susi` will be exported. Additionally, if you execute `kdb set user:/my/app/hello world` the key will not be exported at all, as there are no changes compared to the beginning of the recording.
+
+The same principle applies to the creation and deletion of keys within a session. If you create a key and delete it in the same session, it will not be exported.
+
+By default, the elektra dump format is used. The optional --format parameter can be used to specify which storage plugin should be used to format the output.
+
+## Importing configuration changes
+
+The output of the `kdb record export` command can be imported with the existing `kdb import` functionality. Depending on the specified storage plugin, other steps may be necessary, e.g. the Ansible plugins creates an Ansible playbook.
+
+## Example workflow
+
+The following example changes the IP address of www.google.com, adds an entry for www.microsoft.com and removes www.apple.com. It then shows the changes and outputs an Ansible playbook.
+
+```
+$ kdb record start AddNewHosts
+$ kdb set system:/hosts/ipv4/www.google.com 8.8.8.8
+$ kdb set system:/hosts/ipv4/www.microsoft.com 4.4.4.4
+$ kdb rm system:/hosts/ipv4/www.apple.com
+$ kdb record stop
+$ kdb record changes
+Changed system:/hosts/ipv4/www.google.com from 1.2.3.4 to 8.8.8.8
+Added system:/hosts/ipv4/www.microsoft.com with value 4.4.4.4
+Removed system:/hosts/ipv4/www.apple.com
```

Comment:
Maybe there should also be a way to mark the initial state of a keys as important.

```sh
$ kdb record mark system:/hosts/ipv4/www.libelektra.org
$ kdb record changes
Unchanged system:/hosts/ipv4/www.libelektra.org must have value 7.7.7.7
```

This would ensure that `system:/hosts/ipv4/www.libelektra.org` is recorded and exported with the current value.

In the diff/patch system I mentioned, this could be used to force compatible KDB states. In the export/import system it is less important since you could just change and then change back the key to get it into the changeset. But sometimes there may be keys that can't be changed without side-effects that you don't want during the session recording. For example, you might need to record the port a server uses, to make sure the rest of your session is compatible. But if this server is currently running, changing the port could break things.

Reply:
> Wouldn't someone simply edit the exported file in such cases?

Depends how easy it is to edit the file. But it's also a matter of discoverability. Users need to know that they can add such things to the exported file.

Reply:
I think this feature could come in really handy and it doesn't look like it's going to be too much of a hassle to implement. I'll consider it.

Reply:
Editing the files obviously has a big drawback if you are in a round-trip workflow. So yes, some explicit statement of that you need the default as it is now makes sense.

It would bring quite a number of new cases, though: also marked keys could be changed, added or removed.

Reply:
> It would bring quite a number of new cases, though: also marked keys could be changed, added or removed.

I don't think so. IMO "marking" a key could internally just be implemented as recording a "Changed user:/foo from A to A", i.e. with the `/old` and `/new` keys being equal.

Reply:
@kodebach I don't understand your sentence.

I mean cases like that a marked key could be removed and a later time reappear, then it still should be marked.

Reply:
We could restrict marking keys to keys that have not been modified yet in the current session. Before an import we then check that those marked keys have the exact value. 

Alternatively, we could just "mark" all changed keys implicitly. We already record the old values of modified keys. We could just check that the "old" values of all changed keys that the will be imported equal the values in the target database. This way we don't need an explicit mark command.  Additionally, we can provide a command line paramter like `--dont-check` that skips this.

Reply:
> We could just check that the "old" values of all changed keys that the will be imported equal the values in the target database

I would've done this anyway. If you don't check for the old value, the whole thing isn't really different from just making your modifications and doing a normal `kdb export` and a normal `kdb import` on the other end.

> Additionally, we can provide a command line paramter like `--dont-check` that skips this.

Yes, either that, or there should be away to do this on a key by key level. But the default should the that the old value is checked. Otherwise, you accidentally replace something important.

> This way we don't need an explicit mark command. 

We still do, because maybe you want to mark an unchanged key. You cannot mark all of those, because then you're just doing an export of the whole KDB.

---

I think before we can discuss this properly, we need to distinguish between a _diff_ and a _log_.

```sh
kdb set user:/foo 123
kdb record start

kdb record mark user:/foo
kdb rm user:/foo
kdb set user:/foo 123
```

After this sequence there will be a _log_ of everything that happend somewhere in `user:/elektra/record`. This log will contain something like:

```
Started session
Marked key user:/foo with value 123
Removed key user:/foo with value 123
Added key user:/foo with value 123
```

In this log there is no problem at all. My proposal here was that instead of introducing a new "Marked key ... with value ..." type of log entry, we just let the log look like this (even if it looks a bit silly):

```
Started session
Changed key user:/foo from value 123 to value 123
Removed key user:/foo with value 123
Added key user:/foo with value 123
```

Of course this is just the internal representation. On the user-facing side, we should use a more sensible text.

Now all the problems you are thinking of only arise, when you try to collapse this _log_ into a single _diff_. I don't know whether `user:/foo` should still be marked in this case. IMO it is a matter of definition, if removing a key removes it's marked status. Things become even more complicated, when you also add a final change:

```
Started session
Marked key user:/foo with value 123   |   Changed key user:/foo from value 123 to value 123
Removed key user:/foo with value 123
Added key user:/foo with value 123
Changed key user:/foo from value 123 to value 456
```

Now what? Is the key now marked with value 456? Has it now lost the marking?

This is actually, why I think we really need a way to export the _log_ and not just the _diff_. That's also why I wanted `kdb record replay` (which could also take an exported log as input). The `replay` command would replay the _log_ one step at a time, instead of just applying a _diff_. So in our example a replay would:

1. Check that `user:/foo` has the value `123`
2. Check that `user:/foo` has the value `123` and remove it
3. Add the key `user:/foo` with value `123`
4. (Check that `user:/foo` has value `123` and change it's value to `456`)

That's where the "Changed user:/foo from A to A" idea came from. We need to check that the current value is `A` anyway (otherwise there is a conflict) so the fact that we then set it to `A` doesn't change anything, but it achieved the goal of having a check.

**Final note:** So far this discussion is only about changing one key at a time. But it must of course be possible to have "changed/added/removed/marked" for multiple keys in a single log entry.